### ISSUE TITLE
Install systemd service with debian package

### ISSUE THREAD

**Issue by defuse** [zcash/zcash#1732]
It would be even more awesome if the debian package set up the systemd stuff. Here are the steps I followed to get Zcash launching on boot. Feel free to copy this into a wiki page if it'll be useful to others.

Add a user for zcash to run under:

```
adduser zcash
```

Make the zcash.conf:

```
su zcash
cd ~
mkdir .zcash
vim zcash.conf
```

Generate a random username and password, and create the file like this:

```
rpcuser=<username>
rpcpassword=<password>
gen=1
equihashsolver=tromp
genproclimit=4
port=18233
```

Change `genproclimit` in the above to however many cores you want to devote to mining. If you just want to run a full node, without mining, set `gen=0`.

As root, put this in `/etc/systemd/system/zcash.service`:

```
[Unit]
Description=Zcash Node
After=network.target

[Service]
Type=simple
Restart=always
User=zcash
ExecStart=/usr/bin/zcashd
ExecStop=/usr/bin/zcash-cli stop
PIDFile=/home/zcash/.zcash/zcashd.pid

[Install]
WantedBy=multi-user.target
```

Now start it:

```
systemctl start zcash
```

And enable it so it comes up after a reboot:

```
systemctl enable zcash
```

Open the P2P port to accept incoming connections (and help the network). Add this to the iptables rules (if you're using iptables):

```
-A INPUT -p tcp --dport 18233 -j ACCEPT
```

Now reboot and make sure it all starts correctly. Test from another Internet connection whether that port is accessible.

**Comment by ageis** [comment]
@defuse Great instructions, thanks for supplying them [https://github.com/zcash/zcash/issues/1612](see also).

I use systemd extensively for running zcashd on company servers. I thought about it previously and I personally didn't think it makes sense to include it in the package. The reason is because you never know what local user someone is running as; here you're running it and assuming the user is `zcash`.

If others agree that creating a `zcash` user on the system and enforcing the service to run as it is reasonable, then I could get on board.

**Comment by ITH4Coinomia** [comment]
@defuse Yes, currently I'm running service as "zcash" user, please lemme know the contribution areas I would like to take up this task!

**Comment by ageis** [comment]
What do the other engineers on the team think about this? Should we do it? It also means we're going to have to start shipping a default configuration. We could leave the service disabled by default so it must be manually enabled. We already assume everyone has root and the ability to install packages, so this could ease the burden of starting and getting set up.

The only thing is, they would have to switch users to `zcash` every time they want to run `zcash-cli` and talk to the daemon. That's kind of annoying, and ideally, you don't want `zcash` to have a shell because of the security exposure associated with that. Is there a way to open up the socket to the world?

**Comment by lucaswerkmeister** [comment]
I would also suggest adding some sandboxing options to the service file, for example:

```desktop
NoNewPrivileges=yes
PrivateTmp=yes
PrivateDevices=yes
ProtectSystem=full
ProtectHome=read-only
ReadWritePaths=/home/zcash/.zcash
ReadOnlyPaths=/home/zcash/.zcash/zcash.conf
RestrictAddressFamilies=AF_UNIX AF_INET AF_INET6
MemoryDenyWriteExecute=yes
```
And perhaps also a `SystemCallFilter` (white- or blacklist).

Also, systemd offers a mechanism to set up users for a service (see `sysusers.d(5)`), though I’m not sure if Debian encourages its use or not. (There is also a mechanism for setting up files, `tmpfiles.d(5)`, but I don’t think it can be used it for `zcash.conf` because there’s no way to choose a random username and password.)

**Comment by ageis** [comment]
@lucaswerkmeister Most of those look good, except `MemoryDenyWriteExecute=yes` in my experience is going to cause the service to fail. I ran into a similar issue running zcashd under PaX/grsecurity.

**Comment by lucaswerkmeister** [comment]
Aw, that’s a shame… does the daemon do some JIT code generation or something similar?

**Comment by str4d** [comment]
Note that there's no need to choose a random username and password on setup or first start any longer, as #1999 has been merged.

**Comment by ageis** [comment]
@nathan-at-least Can I solicit your input on whether we should ship this?

**Comment by ageis** [comment]
Let's say we ship a systemd service, a system user named `zcash`, and a default zcash.conf.
At a minimum, in order to use `zcash-cli`, regular users will have to symlink or copy the configuration to their ~/.zcash or supply it as -conf option to access RPC. Imagine if the conf for this service were world-readable at `/etc/zcash.conf`? You also don't want to grant zcash a shell, or require people to switch user in order to do something. I would basically be concerned that any other user on the system could then access the daemon and steal coins. The confinement of user privileges and giving people a choice in that is a nice thing... Granted, if you can run apt-get, then you're a privileged user, but on a multi-user system, you don't want a zcashd running that anyone can access via RPC.

**Comment by mdr0id** [comment]
Note: #2363

**Comment by mms710** [comment]
This is currently blocked on #3147.
